File: typescript/systemdesign/roundRobinLoadBalancer.ts
/*
Round-robin load balancer: cycles through a pool of servers in order,
skipping any currently marked unhealthy, so traffic keeps flowing evenly
around outages.

status - completed
*/

import { logp } from "../utils/logger";

class RoundRobinLoadBalancer {
  private readonly servers: string[];
  private readonly healthy: Set<string>;
  private cursor = 0;

  constructor(servers: string[]) {
    if (servers.length === 0) {
      throw new Error("server pool cannot be empty");
    }
    this.servers = [...servers];
    this.healthy = new Set(servers);
  }

  markUnhealthy(server: string): void {
    this.healthy.delete(server);
  }

  markHealthy(server: string): void {
    if (this.servers.includes(server)) {
      this.healthy.add(server);
    }
  }

  next(): string {
    if (this.healthy.size === 0) {
      throw new Error("no healthy servers available");
    }
    for (let attempts = 0; attempts < this.servers.length; attempts++) {
      const candidate = this.servers[this.cursor];
      this.cursor = (this.cursor + 1) % this.servers.length;
      if (this.healthy.has(candidate)) {
        return candidate;
      }
    }
    throw new Error("no healthy servers available");
  }
}

// usage scenarios

const balancer = new RoundRobinLoadBalancer(["server-a", "server-b", "server-c"]);

const firstRound = Array.from({ length: 6 }, () => balancer.next());
logp(`round robin over 6 requests, all healthy: ${firstRound.join(", ")}`);

balancer.markUnhealthy("server-b");
const secondRound = Array.from({ length: 6 }, () => balancer.next());
logp(`server-b marked unhealthy, next 6 requests skip it: ${secondRound.join(", ")}`);
logp(`server-b appears in that batch: ${secondRound.includes("server-b")} (expected false)`);

balancer.markHealthy("server-b");
const thirdRound = Array.from({ length: 3 }, () => balancer.next());
logp(`server-b marked healthy again, rejoins rotation: ${thirdRound.join(", ")}`);

File: python/systemdesign/round_robin_load_balancer.py
"""
Round-robin load balancer distributing requests evenly across a pool
of servers, skipping any server currently marked unhealthy.

status - completed
"""

from base_logger.logging_event import create_logger


class RoundRobinLoadBalancer:

    log = create_logger(__name__)

    def __init__(self, servers):
        self.servers = list(servers)
        self.healthy = {server: True for server in self.servers}
        self._next_index = 0

    def mark_unhealthy(self, server):
        self.healthy[server] = False

    def mark_healthy(self, server):
        self.healthy[server] = True

    def get_next_server(self):
        total = len(self.servers)
        for _ in range(total):
            server = self.servers[self._next_index]
            self._next_index = (self._next_index + 1) % total
            if self.healthy[server]:
                return server
        return None


if __name__ == "__main__":
    log = create_logger("round_robin_lb_demo")

    lb = RoundRobinLoadBalancer(["s1", "s2", "s3"])

    picks = [lb.get_next_server() for _ in range(6)]
    log.info("6 picks across 3 healthy servers: %s", picks)
    log.info("expected s1,s2,s3,s1,s2,s3 cycling evenly")

    lb.mark_unhealthy("s2")
    picks_with_unhealthy = [lb.get_next_server() for _ in range(4)]
    log.info("s2 marked unhealthy, next 4 picks: %s", picks_with_unhealthy)
    log.info("expected s2 to never appear")

    lb.mark_unhealthy("s1")
    lb.mark_unhealthy("s3")
    log.info("all servers unhealthy -> get_next_server() = %s (expected None)",
             lb.get_next_server())

    lb.mark_healthy("s2")
    log.info("s2 recovers -> get_next_server() = %s", lb.get_next_server())
